src/java/org/apache/cassandra/tools/profiler/AsyncProfilerService.java

```diff
@@ -0,0 +1,166 @@
+
+    private static AsyncProfiler profilerInstance;
+
+    static
+    {
+        try
+        {
+            // Let async-profiler automatically extract and load the native library from the JAR
+            profilerInstance = AsyncProfiler.getInstance();
+        }
+        catch (Throwable t)
+        {
+            logger.error("async-profiler initialization ERROR");
+            profilerInstance = null;
+        }
+    }
+
+    public void start(String event, String outputFormat, int timeout, String outputFileName)
+    {
+        checkProfilerInstance();
+        validateEvent(event);
+        validateFormat(outputFormat);
+        validateOutputFileName(outputFileName);
+        String outputPath = Path.of(ASYNC_PROFILER_OUTPUT_DIRECTORY.getString(), outputFileName).toString();
+        try
+        {
+            String cmd = String.format("start,%s,event=%s,timeout=%s,file=%s",
+                                       outputFormat,
+                                       event,
+                                       timeout,
+                                       outputPath);
+
+            profilerInstance.execute(cmd);
+            logger.info("Started async-profiler: cmd={}", cmd);
+        }
+        catch (IOException e)
+        {
+            logger.error("Failed to start async-profiler", e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    public void stop(String outputFileName)
+    {
+        checkProfilerInstance();
+        validateOutputFileName(outputFileName);
+
+        String outputPath = Path.of(ASYNC_PROFILER_OUTPUT_DIRECTORY.getString(), outputFileName).toString();
+
+        String cmd = String.format("stop,file=%s",
+                                   outputPath);
+
+        try
+        {
+            profilerInstance.execute(cmd);
+            logger.info("Stopped async-profiler: cmd={}", cmd);
+        } catch (IOException e)
+        {
+            logger.error("Failed to stop async-profiler", e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    public void execute(String command)
+    {
+        checkProfilerInstance();
+
+        try
+        {
+            profilerInstance.execute(command);
+            logger.info("Executed raw async-profiler command {}", command);
+        }
+        catch (IOException e)
+        {
+            logger.error("Failed to execute raw async-profiler command {}", command, e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    public boolean isAvailable()
+    {
+        return profilerInstance != null;
+    }
+
+    private void checkProfilerInstance()
+    {
+        if (ASYNC_PROFILER_ENABLED.getBoolean() == false)
+        {
+            throw new IllegalStateException("async-profiler is not enabled.");
+        }
+        else if (!isAvailable())
+        {
+            throw new IllegalStateException("async-profiler is not initialized.");
+        }
+    }
+
+    private void validateEvent(String event)
```

Comment:
These validation methods might be replaced by a static method on each respective logical enum which you can reuse in nodetool.